File: intellij-ides/src/main/kotlin/com/ant/services/OpenAITranslationService.kt
package com.ant.services

import com.ant.utils.LoggerUtils.getLogger
import com.google.gson.Gson
import com.google.gson.JsonObject
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

/**
 * OpenAI API service for translation
 */
class OpenAITranslationService(private val apiKey: String) {
    private val logger = getLogger(javaClass)
    private val httpClient = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(30))
        .build()
    private val gson = Gson()
    
    /**
     * Translate text using OpenAI GPT-4
     */
    fun translateText(text: String, sourceLanguage: String, targetLanguage: String): String? {
        try {
            if (apiKey.isBlank()) {
                logger.warn("OpenAI API key is not configured")
                return null
            }
            
            val prompt = """
                Translate the following text from $sourceLanguage to $targetLanguage.
                Only return the translated text, nothing else.
                
                Text to translate: "$text"
            """.trimIndent()
            
            val requestBody = JsonObject().apply {
                addProperty("model", "gpt-4")
                add("messages", gson.toJsonTree(listOf(
                    mapOf("role" to "user", "content" to prompt)
                )))
                addProperty("max_tokens", 150)
                addProperty("temperature", 0.3)
            }
            
            val request = HttpRequest.newBuilder()
                .uri(URI.create("https://api.openai.com/v1/chat/completions"))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer $apiKey")
                .POST(HttpRequest.BodyPublishers.ofString(gson.toJson(requestBody)))
                .timeout(Duration.ofSeconds(30))
                .build()
            
            val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
            
            if (response.statusCode() == 200) {
                val responseJson = gson.fromJson(response.body(), JsonObject::class.java)
                val choices = responseJson.getAsJsonArray("choices")
                if (choices.size() > 0) {
                    val message = choices[0].asJsonObject
                        .getAsJsonObject("message")
                    val rawTranslatedText = message.get("content").asString.trim()
                    
                    // Remove surrounding quotes if present
                    val translatedText = rawTranslatedText
                        .trim()
                        .removePrefix("\"")
                        .removeSuffix("\"")
                        .removePrefix("'")
                        .removeSuffix("'")
                        .trim()
                    
                    logger.info("Translation successful: '$text' -> '$translatedText' ($sourceLanguage -> $targetLanguage)")
                    return translatedText
                }
            } else {
                logger.warn("OpenAI API error: ${response.statusCode()} - ${response.body()}")
            }
        } catch (e: Exception) {
            logger.error("Error during translation: ${e.message}", e)
        }
        
        return null
    }
    
    /**
     * Test API key validity
     */
    fun testApiKey(): Boolean {
        return try {
            val testTranslation = translateText("Hello", "English", "Spanish")
            testTranslation != null
        } catch (e: Exception) {
            logger.warn("API key test failed: ${e.message}")
            false
        }
    }
} 

File: visual-studio-code/src/openaiTranslationService.ts
import axios from 'axios';

export class OpenAITranslationService {
    private readonly baseUrl = 'https://api.openai.com/v1/chat/completions';

    /**
     * Translate text using OpenAI GPT-4
     */
    async translateText(
        text: string,
        sourceLanguage: string,
        targetLanguage: string,
        apiKey: string
    ): Promise<string | null> {
        if (!apiKey || !text.trim()) {
            return null;
        }

        return await this.translateWithRetry(text, sourceLanguage, targetLanguage, apiKey, 3);
    }

    /**
     * Translate with retry mechanism
     */
    private async translateWithRetry(
        text: string,
        sourceLanguage: string,
        targetLanguage: string,
        apiKey: string,
        maxRetries: number
    ): Promise<string | null> {
        for (let attempt = 1; attempt <= maxRetries; attempt++) {
            try {
                const prompt = this.createTranslationPrompt(text, sourceLanguage, targetLanguage);
                
                const response = await axios.post(
                    this.baseUrl,
                    {
                        model: 'gpt-4',
                        messages: [
                                                    {
                            role: 'system',
                            content: 'You are a professional translator. Translate accurately while preserving the EXACT formatting of the source text. CRITICAL: Match the exact capitalization pattern and punctuation of the source. If source starts lowercase, translation must start lowercase. If source has no period, translation must have no period. Return ONLY the translated text without quotes or additional formatting.'
                        },
                            {
                                role: 'user',
                                content: prompt
                            }
                        ],
                        max_tokens: 1000,
                        temperature: 0.3
                    },
                    {
                        headers: {
                            'Authorization': `Bearer ${apiKey}`,
                            'Content-Type': 'application/json'
                        },
                        timeout: 30000
                    }
                );

                let translatedText = response.data.choices[0]?.message?.content?.trim();
                
                // Clean up the response - remove quotes and escape characters
                if (translatedText) {
                    // Remove surrounding quotes if present
                    if ((translatedText.startsWith('"') && translatedText.endsWith('"')) ||
                        (translatedText.startsWith("'") && translatedText.endsWith("'"))) {
                        translatedText = translatedText.slice(1, -1);
                    }
                    
                    // Remove escape characters
                    translatedText = translatedText.replace(/\\"/g, '"').replace(/\\'/g, "'");
                }
                
                return translatedText || null;

            } catch (error) {
                console.error(`OpenAI translation error (attempt ${attempt}/${maxRetries}):`, error);
                
                if (axios.isAxiosError(error)) {
                    const status = error.response?.status;
                    
                    // Don't retry for these errors
                    if (status === 401) {
                        throw new Error('Invalid OpenAI API key');
                    }
                    
                    // Retry for these errors
                    if (status === 429 || status === 500 || status === 502 || status === 503 || status === 504) {
                        if (attempt < maxRetries) {
                            const delay = Math.min(1000 * Math.pow(2, attempt - 1), 10000); // Exponential backoff, max 10s
                            console.log(`Retrying in ${delay}ms... (attempt ${attempt + 1}/${maxRetries})`);
                            await this.sleep(delay);
                            continue;
                        } else {
                            throw new Error(`OpenAI API temporarily unavailable (${status}). Please try again later.`);
                        }
                    }
                    
                    throw new Error(`OpenAI API error: ${status || 'Unknown'}`);
                }
                
                // For non-axios errors, don't retry
                throw new Error(`Translation failed: ${error}`);
            }
        }
        
        return null;
    }

    /**
     * Sleep utility for retry delays
     */
    private sleep(ms: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    /**
     * Create translation prompt
     */
    private createTranslationPrompt(text: string, sourceLanguage: string, targetLanguage: string): string {
        const sourceLanguageName = this.getLanguageName(sourceLanguage);
        const targetLanguageName = this.getLanguageName(targetLanguage);
        
        return `Translate the following text from ${sourceLanguageName} to ${targetLanguageName}:

${text}

CRITICAL formatting rules:
- Preserve EXACT capitalization pattern of the source text (if source starts lowercase, translation must start lowercase)
- Preserve EXACT punctuation of the source text (if source has no period, translation must have no period)
- Keep the same tone and style as the source
- If it's a UI text, make it natural for the target language while maintaining formatting
- Return ONLY the translated text without quotes or any additional formatting
- Do not add punctuation that doesn't exist in the source
- Do not change capitalization from the source pattern`;
    }

    /**
     * Get language name from language code
     */
    private getLanguageName(languageCode: string): string {
        const languageNames: { [key: string]: string } = {
            'en': 'English',
            'tr': 'Turkish',
            'es': 'Spanish',
            'fr': 'French',
            'de': 'German',
            'it': 'Italian',
            'pt': 'Portuguese',
            'ru': 'Russian',
            'ja': 'Japanese',
            'ko': 'Korean',
            'zh': 'Chinese',
            'ar': 'Arabic',
            'hi': 'Hindi',
            'nl': 'Dutch',
            'sv': 'Swedish',
            'da': 'Danish',
            'no': 'Norwegian',
            'fi': 'Finnish',
            'pl': 'Polish',
            'cs': 'Czech',
            'sk': 'Slovak',
            'hu': 'Hungarian',
            'ro': 'Romanian',
            'bg': 'Bulgarian',
            'hr': 'Croatian',
            'sr': 'Serbian',
            'sl': 'Slovenian',
            'et': 'Estonian',
            'lv': 'Latvian',
            'lt': 'Lithuanian',
            'uk': 'Ukrainian',
            'be': 'Belarusian',
            'mk': 'Macedonian',
            'mt': 'Maltese',
            'is': 'Icelandic',
            'ga': 'Irish',
            'cy': 'Welsh',
            'eu': 'Basque',
            'ca': 'Catalan',
            'gl': 'Galician',
            'pt-br': 'Brazilian Portuguese',
            'zh-cn': 'Simplified Chinese',
            'zh-tw': 'Traditional Chinese'
        };

        return languageNames[languageCode.toLowerCase()] || languageCode.toUpperCase();
    }

    /**
     * Test API key validity
     */
    async testApiKey(apiKey: string): Promise<boolean> {
        if (!apiKey) {
            return false;
        }

        try {
            const response = await axios.post(
                this.baseUrl,
                {
                    model: 'gpt-3.5-turbo',
                    messages: [
                        {
                            role: 'user',
                            content: 'Hello'
                        }
                    ],
                    max_tokens: 5
                },
                {
                    headers: {
                        'Authorization': `Bearer ${apiKey}`,
                        'Content-Type': 'application/json'
                    },
                    timeout: 10000
                }
            );

            return response.status === 200;
        } catch {
            return false;
        }
    }
} 